design/ceph/object/store.md

```diff
@@ -712,3 +712,64 @@ For external CephCephObjectStores (i.e., when `spec.gateway.externalRgwEndpoints
+store. Any user configs that break the object store should be a responsibility of the user.
+
+```yaml
+spec:
+  metadataPool: # ...
+  dataPool: # ...
+  gateway: # ...
+
+  cephConfig:
+    rgw_enable_apis: s3, s3website, swift, swift_auth, sts, iam, notifications # e.g. 'admin' removed
+    rgw_gc_max_concurrent_io: "4" # e.g, reduce garbage collection IO
+    debug_rgw: "20"
+```
+
+CephCluster has a similar configuration section called `config`. `config` may be somewhat too vague
+in the context of the CephObjectStore CR. `cephConfig` is a clearer naming for users and give
+context that Ceph configs are expected.
+
+Rook will apply user config values to the mon config store. Rook will not override its own mon
+config values set via other means. This will ensure that values controlled by Rook for other spec
+configurations or that are important for running in Kubernetes are not clobbered by user configs.
+This will risk some instances where users want to update a value that Rook currently sets.
+This tradeoff favors reliability over flexibility.
+
+Examples:
+    - Rook sets configs like `rgw_zone` which are critical for a functional CephObjectStore. Users
+    would not be able to override this value.
+    - Rook sets `rgw_dns_name` via the `hosting` config section. User inputs to `rgw_dns_name`
+    wouldn't take affect when `hosting` configs are set.
+
+If possible, Rook should make log when it overrides a user's `cephConfig` based on another config or
+internal decision. This may not be easily feasible when Rook applies CLI flags to the RGW
+deployment.
+
+Rook will apply each configuration under the client scope of the shared RGW client.
+i.e., `client.rook-ceph-rgw-<store-name>`. This will also help ensure that configs changed don't
+affect other Rook-managed Ceph daemons, critically mons, osds, and mds.
+
+Rook has the option to validate that each config is prefixed with `rgw_`. This would prevent users
+from modifying configs that are unrelated to RGW functionality. For example, users would be
+prevented from setting a config like `admin_socket` which might affect RGW's ability to run
+correctly in the container. As a downside, this would also mean that users are unable to set
+non-affecting configs like `debug_rgw`.
+[Ceph general configs](https://docs.ceph.com/en/reef/rados/configuration/general-config-ref/) that
+apply to all daemons are unlikely to be used by non-advanced users and probably unlikely to result
+in issues keeping RGWs running. Based on the perceived low risk, and the definite benefit for users
+especially for debugging the object store, Rook will not validate configs.
+
+This config does not allow users to override CLI flags on the RGW deployment; only mon config store
+values. This means that users may need to restart the RGW manually themselves for certain
+configurations. Rook code could allow users to override CLI flags to allow automatic restarts, but
+this would raise greater risk of
```

Comment:
@travisn @arttor do you have thoughts on whether we (now or in the future) want to allow some config changes to cause an RGW restart automatically? The way the API is proposed currently `spec.config.map[string]string`, there isn't much flexibility for changing the API in the future to allow setting values that require automatic restart

Reply:
Right. I considered that we could allow users to specify additional CLI flags, but that seems much more risky than mon kv changes

Reply:
i think that it would be useful to restart rgw automatically if it is required to apply the option. Probably, we can split this config into 2 maps:  `spec.initConfig` and `spec.runtimeConfig`. And then set options as args from the first map and to mon db from the second.

But what do you think about setting all options as args to restart rgw on every change? Since RGW is stateless it should be safe. And in this way it will be easy to observe current configs by looking at k8s manifests. Otherwise, users will have to install toolbox and use `ceph config dump` which is not always possible.

Another interesting topic is secret options. Is it makes sense to support it with pod envars referring k8s secrets or by mounting secret into Pod? With secret options we can cover some integration features like keystone (already implemented in CRD) or [Barbican](https://docs.ceph.com/en/reef/radosgw/config-ref/#barbican-settings) (not supported https://github.com/rook/rook/issues/14830). In theory, such features can be fully configured with presented config and will not requre to introduce CRD and controller code  changes.

Reply:
Secrets can be mounted into files today via this mechanism, which I added a while back fairly quietly: https://github.com/rook/rook/pull/14616

I can see some benefit to potentially adding env vars, but let's leave that for a future enhancement. 

I like your idea of having `initConfig` and `runtimeConfig` to distinguish restart/non-restart behavior.

Reply:
I was not aware of custom volume mounts but it looks pretty good for setting secret options. If i got it right, such config defined in local file will be local to the rgw daemon and won't affect other rgw instances in case of multi-instance rgw deployment. Do we need to add `[client.radosgw.{instance-name}]` to mountied config in this case or it will work if we just put all options under `[global]`?